src/metabase/models/secret.clj

```diff
@@ -100,6 +101,22 @@
           (.write out v)))
       tmp-file)))
 
+(def ^{:arglists '([{:keys [connection-property-name id value] :as secret} driver?])}
+  value->file!
+  "Returns the value of the given `secret` instance in the form of a file. If the given instance has a `:file-path` as
+  its source, a `File` referring to that is returned. Otherwise, the `:value` is written to a temporary file, which is
+  then returned.
+
+  `driver?` is an optional argument that is only used if an ostensibly existing file value (i.e. `:file-path`) can't be
+  resolved, in order to render a more user-friendly error message (by looking up the display names of the connection
+  properties involved)."
+  (memoize/memo
+   (with-meta value->file!*
+     {::memoize/args-fn (fn [[secret _driver?]]
+                          ;; not clear if value->string could return nil due to the cond so we'll just cache on a key
+                          ;; that is unique
+                          [(or (value->string secret) (gensym))])})))
```

Comment:
Wouldn't `value->string` throw an exception if the byte array is not a valid UTF-8 string?

Reply:
good thought. I'll catch and use the gensym there as well

Reply:
(although I'm assuming this code path doesn't throw a lot)

Reply:
Actually it seems `String.` doesn't care about valid bytes. Makes me think we don't need to catch here at all. We just need a stable key and goofy non-printing strings are fine?
<img width="414" alt="image" src="https://user-images.githubusercontent.com/6377293/181120053-3d2b7c1c-8131-4165-b559-f61c24402b8c.png">

Reply:
I don't think it's very important how it prints, but the javadoc says this about invalid encodings:

> The behavior of this constructor when the given bytes are not valid in the given charset is unspecified. The [CharsetDecoder](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/charset/CharsetDecoder.html) class should be used when more control over the decoding process is required.

So for maximum safety you could use that or just [hash](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Arrays.html#hashCode(byte%5B%5D)) the byte array.

Reply:
can hash the byte array:
```clojure
secret=> (hash (.getBytes "hi"))
917686548
secret=> (hash (.getBytes "hi"))
718722076
```
unless we have a hash that works on the contents:
```clojure
secret=> (java.util.Arrays/hashCode ^bytes (.getBytes "hi"))
4290
secret=> (java.util.Arrays/hashCode ^bytes (.getBytes "hi"))
4290
```
But this leaves us pretty susceptible to hash collisions so i'm nervous about that.

Reply:
One thing I could do is just turn the byte array into a clojure vector and use that. That has the semantics we want I think: value based rather than identity, collisions only occur if and only if the sequence of bytes are equal

Reply:
I don't understand what makes you nervous about `Arrays/hashCode`? Do you expect that to cause more collisions than hashing the equivalent `String`?

Reply:
I think I made this confusing. When you upload a sequence of bytes as a secret, we want to put them in a file once and only once and always reuse that temporary file. We will eventually hash the whole connection spec but I don't care about collisions there. It's only given the same sequence of bytes, you should always get back the exact same temporary file that has been created.

If the memoization key is the byte-array itself, this will fail because arrays have reference identity:
```clojure
user=> (= (.getBytes "hi") (.getBytes "hi"))
false
```
(I was using the fact that they had different hashes as a shorthand for this: their notion of identity is tied to where in memory those byte arrays live rather than their contents).

This means that memoization cannot be driven off of the byte array. But one way to gain this back quickly is just stuff those bytes into a string, and then strings compare on value not identity. This is what is currently in the PR. I was banking on the assumption that Strings are just opaque sequences of bytes that will compare byte by byte, regardless of whether those bytes make sense.

But you've pointed out a good point that maybe that is flirting with undefined behavior. If we use the hash of the contents of the byte array as the memoization key with `(j.u.Arrays/hashCode array)`, then we open ourselves up the (albeit rare) case that two distinct secret values hash to the same value. This sounds really bad.

An easy way to have the value semantics we want for the memoization is just to call `(vec array)` on the byte array and use this sequence of bytes as the memoization key. Clojure vectors compare by value not reference. So two secrets would return the same file if and only if the sequence of bytes are identical, in which case we would expect the files to be identical. This gives me the same guarantee that I was wanting from the `String` behavior I used initially without entwining this with charsets, utf8, etc.

Reply:
Yes, when I managed to defocus my brain from hashing, I understood this. 😄